src/OrchardCore.Modules/OrchardCore.Templates/Controllers/PreviewController.cs

```diff
@@ -1,15 +1,89 @@
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Mvc;
-using OrchardCore.Admin;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Display;
 using OrchardCore.DisplayManagement.ModelBinding;
+using OrchardCore.Modules;
+using OrchardCore.Settings;
+using OrchardCore.Templates.ViewModels;
 
 namespace OrchardCore.Templates.Controllers
 {
-    [Admin]
     public class PreviewController : Controller, IUpdateModel
     {
+        private readonly IContentManager _contentManager;
+        private readonly IContentAliasManager _contentAliasManager;
+        private readonly IContentItemDisplayManager _contentItemDisplayManager;
+        private readonly IAuthorizationService _authorizationService;
+        private readonly ISiteService _siteService;
+        private readonly IClock _clock;
+
+        public PreviewController(
+            IContentManager contentManager,
+            IContentAliasManager contentAliasManager,
+            IContentItemDisplayManager contentItemDisplayManager,
+            IAuthorizationService authorizationService,
+            ISiteService siteService,
+            IClock clock)
+        {
+            _contentManager = contentManager;
+            _contentAliasManager = contentAliasManager;
+            _contentItemDisplayManager = contentItemDisplayManager;
+            _authorizationService = authorizationService;
+            _siteService = siteService;
+            _clock = clock;
+        }
+
         public IActionResult Index()
         {
             return View();
         }
+
+        [HttpPost]
+        public async Task<IActionResult> Render()
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.ManageTemplates))
+            {
+                return Unauthorized();
+            }
+
+            var name = Request.Form["Name"];
+            var content = Request.Form["Content"];
+
+            if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(content))
+            {
+                HttpContext.Items["OrchardCore.PreviewTemplate"] = new TemplateViewModel { Name = name, Content = content };
+            }
+
+            var alias = Request.Form["Alias"].ToString();
+
+            string contentItemId;
+            if (string.IsNullOrEmpty(alias) || alias == "/" || alias.EndsWith("Preview/Index"))
+            {
+                var homeRoute = _siteService.GetSiteSettingsAsync().GetAwaiter().GetResult().HomeRoute;
+                contentItemId = homeRoute["contentItemId"]?.ToString();
+            }
+            else
+            {
+                contentItemId = await _contentAliasManager.GetContentItemIdAsync("slug:" + alias);
```

Comment:
What's the reason to do this?

Reply:
While previewing a template we send the current frame location path (notice that we can navigate in the preview window), so we can retrieve a related content item id and then implement template previewing quite the same way as content item previewing. I mean by using the returned model data in place of always reloading the frame with a new location.